### PR TITLE
Make `circuit._prev_moment_available` a public method.

### PR DESCRIPTION
**Pull request by tanujkhattar**
Fixes https://github.com/quantumlib/Cirq/issues/4976

### CODE REVIEW

File: cirq-core/cirq/circuits/circuit.py [hunk 1]
```diff
@@ -1950,7 +1951,27 @@ def transform_qubits(
         with _compat.block_overlapping_deprecation(re.escape(_DEVICE_DEP_MESSAGE)):
             return Circuit(op_list, device=self._device if new_device is None else new_device)
 
-    def _prev_moment_available(self, op: 'cirq.Operation', end_moment_index: int) -> Optional[int]:
+    def prev_moment_available(
+        self, op: 'cirq.Operation', end_moment_index: Optional[int] = None
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply cirq-core/cirq/circuits/circuit.py]: done

**Comment by daxfohl** [comment]
Consider bikeshedding the name if we make it public. Every time I stumble into this method I think it's a boolean. The arg name is a bit odd too.

File: cirq-core/cirq/circuits/circuit.py [hunk 2]
```diff
@@ -1950,7 +1951,27 @@ def transform_qubits(
...
+        that depend on the results of those measurements. Therefore, using this method, a CCO
+        `op` would not be allowed to move left past a measurement it depends upon.
+
+        Args:
+            op: Operation for which the previous moment that can accommodate it needs to be found.
+            end_moment_index: The moment index just after the starting point of the reverse search.
+                Defaults to the length of the list of moments.
+
+        Returns:
+            Index of the latest matching moment. Returns `end_moment_index` if no moment on left is
```
Comment (95-martin-orion) [inline cirq-core/cirq/circuits/circuit.py:1970]: This should be _earliest_ matching moment, right?

File: cirq-core/cirq/circuits/circuit.py [hunk 3]
```diff
@@ -1950,7 +1951,27 @@ def transform_qubits(
         with _compat.block_overlapping_deprecation(re.escape(_DEVICE_DEP_MESSAGE)):
             return Circuit(op_list, device=self._device if new_device is None else new_device)
 
-    def _prev_moment_available(self, op: 'cirq.Operation', end_moment_index: int) -> Optional[int]:
+    def prev_moment_available(
+        self, op: 'cirq.Operation', *, end_moment_index: Optional[int] = None
+    ) -> int:
+        """Finds the index of the previous (i.e. left most) moment which can accommodate `op`.
```
Comment (95-martin-orion) [inline cirq-core/cirq/circuits/circuit.py:1957]: As Dax said, the name doesn't really suggest leftmost - I read it as "get the previous available moment", which seems to suggest _rightmost_. Maybe something like `earliest_equivalent_moment` instead?

File: cirq-core/cirq/circuits/circuit.py [hunk 4]
```diff
@@ -1968,6 +1989,8 @@ def _prev_moment_available(self, op: 'cirq.Operation', end_moment_index: int) ->
             ):
                 return last_available
             if self._can_add_op_at(k, op):
+                # Note: Remove the if condition after `self._device` is gone and move the method to
```
Comment (95-martin-orion) [inline cirq-core/cirq/circuits/circuit.py:1992]: Not sure I understand this comment - `cirq.AbstractDevice` is for post-processing of the circuit, right?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/circuits/circuit.py (hunk 4 above)
Comment (tanujkhattar) [reply cirq-core/cirq/circuits/circuit.py:1992]: `_can_add_op_at` will be moot here once `self._device` is being deprecated -- i.e. circuits will stop using devices for validation, which is being done as part of the devices 2.0 migration (see https://github.com/quantumlib/Cirq/issues/4744)

File: cirq-core/cirq/circuits/circuit.py (hunk 3 above)
Comment (tanujkhattar) [reply cirq-core/cirq/circuits/circuit.py:1957]: I think the word `equivalent` can also cause confusion. How about `earliest_available_moment` ?

File: cirq-core/cirq/circuits/circuit.py (hunk 2 above)
Comment (tanujkhattar) [reply cirq-core/cirq/circuits/circuit.py:1970]: done

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**